File: src/frontend/shared/src/mocks/handlers/work-executions.ts
import { v7 as uuidv7 } from 'uuid';
import {
  HttpResponse,
  envelope,
  problem,
  requireAuth,
  route,
  storeIdempotency,
  withIdempotency,
} from '../_helpers';
import { db } from '../db/seed';
import { computeContentHash, GENESIS_HASH } from '../../domain/hash-chain';
import type {
  CompleteWorkExecutionRequest,
  CreateWorkExecutionRequest,
  ResumeWorkExecutionRequest,
  SuspendWorkExecutionRequest,
  WorkExecution,
} from '../../types';

type CompleteRequest = { completed_by?: string; timestamp_client?: string; final_remarks?: string } & CompleteWorkExecutionRequest;
type SuspendRequest = { reason_code?: string; reason_detail?: string; timestamp_client?: string } & SuspendWorkExecutionRequest;
type ResumeRequest = { resumed_by?: string; timestamp_client?: string } & ResumeWorkExecutionRequest;
type CreateRequest = { work_order_id?: string; operator_id?: string; device_id?: string; start_timestamp_client?: string } & CreateWorkExecutionRequest;

export const workExecutionHandlers = [
  ...route('post', 'terminal', '/work-executions', async ({ request }) => {
    const authErr = requireAuth(request);
    if (authErr) return authErr;
    const body = (await request.json().catch(() => null)) as CreateRequest | null;
    if (!body?.work_order_id || !body.operator_id || !body.device_id) {
      return problem(422, 'ERR-VAL-001', 'Required field missing', '必須フィールドが不足しています');
    }
    const workOrder = db.workOrders.find((wo) => wo.id === body.work_order_id);
    if (!workOrder) {
      return problem(409, 'ERR-DB-002', 'Foreign key violation', 'work_order が存在しません');
    }
    if (db.workExecutions.some((we) => we.workOrderId === body.work_order_id && we.status === 'in_progress')) {
      return problem(409, 'ERR-BIZ-007', 'version_already_published', '同一作業指示で実行中の作業が存在します');
    }
    const idem = await withIdempotency<WorkExecution>(request, body);
    if (idem.conflict) return idem.conflict;
    if (idem.cached) return HttpResponse.json(envelope(idem.cached.response), { status: idem.cached.status });

    const steps = db.steps.filter((s) => s.sopVersionId === workOrder.sopVersionId);
    const exec: WorkExecution = {
      id: uuidv7(),
      workOrderId: workOrder.id,
      operatorId: body.operator_id,
      deviceId: body.device_id,
      status: 'in_progress',
      currentStepId: steps[0]?.id ?? null,
      completedStepCount: 0,
      totalStepCount: steps.length,
      sopVersionSnapshot: {
        sopId: workOrder.sopId,
        version: '1.0.0',
        snapshotHash: 'sha256:mock',
      },
      startedAt: new Date().toISOString(),
      lastEventAt: new Date().toISOString(),
      completedAt: null,
      createdAt: new Date().toISOString(),
    };
    db.workExecutions.push(exec);
    storeIdempotency(idem.key, idem.bodyHash, exec, 201);
    return HttpResponse.json(envelope(exec), { status: 201 });
  }),

  ...route('put', 'terminal', '/work-executions/:id/heartbeat', ({ request, params }) => {
    const authErr = requireAuth(request);
    if (authErr) return authErr;
    const exec = db.workExecutions.find((we) => we.id === params['id']);
    if (!exec) return problem(404, 'ERR-DB-002', 'NotFound', 'work_execution が存在しません');
    exec.lastEventAt = new Date().toISOString();
    return HttpResponse.json(envelope({ id: exec.id, status: exec.status, last_event_at: exec.lastEventAt }));
  }),

  ...route('post', 'terminal', '/work-executions/:id/suspend', async ({ request, params }) => {
    const authErr = requireAuth(request);
    if (authErr) return authErr;
    const exec = db.workExecutions.find((we) => we.id === params['id']);
    if (!exec) return problem(404, 'ERR-DB-002', 'NotFound', 'work_execution が存在しません');
    if (exec.status !== 'in_progress') {
      return problem(409, 'ERR-BIZ-001', 'lock_step_violation', '実行中以外の作業は中断できません');
    }
    const body = (await request.json().catch(() => null)) as SuspendRequest | null;
    if (!body?.reason_code) {
      return problem(422, 'ERR-VAL-001', 'Required field missing', 'reason_code は必須です');
    }
    const suspendedAt = new Date().toISOString();
    const suspensionId = uuidv7();
    exec.status = 'suspended';
    db.suspensions.push({
      id: suspensionId,
      workExecutionId: exec.id,
      reasonCode: body.reason_code as 'equipment_breakdown',
      reasonDetail: body.reason_detail ?? '',
      suspendedAt,
      resumedAt: null,
      resumedBy: null,
    });
    return HttpResponse.json(envelope({
      id: exec.id,
      status: 'suspended',
      suspension_id: suspensionId,
      suspended_at: suspendedAt,
    }));
  }),

  ...route('post', 'terminal', '/work-executions/:id/resume', async ({ request, params }) => {
    const authErr = requireAuth(request);
    if (authErr) return authErr;
    const exec = db.workExecutions.find((we) => we.id === params['id']);
    if (!exec) return problem(404, 'ERR-DB-002', 'NotFound', 'work_execution が存在しません');
    const body = (await request.json().catch(() => null)) as ResumeRequest | null;
    if (!body?.resumed_by) {
      return problem(422, 'ERR-VAL-001', 'Required field missing', 'resumed_by は必須です');
    }
    exec.status = 'in_progress';
    const suspension = db.suspensions.find((s) => s.workExecutionId === exec.id && s.resumedAt === null);
    if (suspension) {
      suspension.resumedAt = new Date().toISOString();
      suspension.resumedBy = body.resumed_by;
    }
    return HttpResponse.json(envelope({
      id: exec.id,
      status: 'in_progress',
      resumed_at: new Date().toISOString(),
      current_step_id: exec.currentStepId,
    }));
  }),

  ...route('post', 'terminal', '/work-executions/:id/complete', async ({ request, params }) => {
    const authErr = requireAuth(request);
    if (authErr) return authErr;
    const exec = db.workExecutions.find((we) => we.id === params['id']);
    if (!exec) return problem(404, 'ERR-DB-002', 'NotFound', 'work_execution が存在しません');
    const body = (await request.json().catch(() => null)) as CompleteRequest | null;
    if (!body?.completed_by) {
      return problem(422, 'ERR-VAL-001', 'Required field missing', 'completed_by は必須です');
    }
    const idem = await withIdempotency<unknown>(request, body);
    if (idem.conflict) return idem.conflict;

    const completedAt = new Date().toISOString();
    exec.status = 'completed';
    exec.completedAt = completedAt;
    const lastBlock = db.hashChainBlocks[db.hashChainBlocks.length - 1];
    const prevHash = lastBlock?.contentHash ?? GENESIS_HASH;
    const blockPayload = { workExecutionId: exec.id, completedAt };
    const contentHash = computeContentHash(prevHash, blockPayload);
    const block = {
      id: uuidv7(),
      blockNumber: db.hashChainBlocks.length + 1,
      prevHash,
      contentHash,
      payload: JSON.stringify(blockPayload),
      createdAt: completedAt,
    };
    db.hashChainBlocks.push(block);
    const response = {
      id: exec.id,
      status: 'completed' as const,
      completed_at: completedAt,
      hash_chain_block_id: block.id,
      hash_chain_value: `sha256:${block.contentHash}`,
    };
    storeIdempotency(idem.key, idem.bodyHash, response, 200);
    return HttpResponse.json(envelope(response));
  }),
];

File: src/backend/crates/wnav_terminal_api/src/dto/work_executions.rs
// 作業実行 API（API-work-execs-001〜005）の DTO 定義（03_作業実行API仕様.md §3〜7）

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;

/// SOP バージョンスナップショット情報
#[derive(Debug, Serialize, ToSchema)]
pub struct SopVersionSnapshot {
    pub sop_id: Uuid,
    pub version: String,
    pub snapshot_hash: String,
}

/// 作業実行イベントサマリ（GET レスポンス内の events 配列要素）
#[derive(Debug, Serialize, ToSchema)]
pub struct WorkEventSummary {
    pub event_id: Uuid,
    pub activity: String,
    pub step_id: Option<Uuid>,
    pub step_number: Option<i32>,
    pub timestamp_server: DateTime<Utc>,
}

// ─────────────────────────────────────────────────────────────────────────────
// API-work-execs-001: POST /api/v1/work-executions
// ─────────────────────────────────────────────────────────────────────────────

/// 作業開始リクエスト（API-work-execs-001）
#[derive(Debug, Deserialize, ToSchema)]
#[allow(dead_code)]
pub struct StartWorkRequest {
    /// 対象作業指示 ID（TBL-006 に open ステータスで存在すること）
    pub work_order_id: Uuid,
    /// 実行オペレータ ID（TBL-016 に存在し、必要スキルを保有すること）
    pub operator_id: Uuid,
    /// 実行端末 ID（TBL-033 に存在すること）
    pub device_id: Uuid,
    /// クライアント側の開始時刻（監査用）
    pub start_timestamp_client: DateTime<Utc>,
}

/// 作業開始レスポンスの data フィールド（API-work-execs-001）
#[derive(Debug, Serialize, ToSchema)]
pub struct StartWorkData {
    pub id: Uuid,
    pub work_order_id: Uuid,
    pub operator_id: Uuid,
    pub device_id: Uuid,
    pub status: String,
    pub current_step_id: Option<Uuid>,
    pub sop_version_snapshot: SopVersionSnapshot,
    pub started_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
}

// ─────────────────────────────────────────────────────────────────────────────
// API-work-execs-002: GET /api/v1/work-executions/{id}
// ─────────────────────────────────────────────────────────────────────────────

/// 作業実行詳細レスポンスの data フィールド（API-work-execs-002）
#[derive(Debug, Serialize, ToSchema)]
pub struct WorkExecutionDetailData {
    pub id: Uuid,
    pub work_order_id: Uuid,
    pub operator_id: Uuid,
    pub device_id: Uuid,
    pub status: String,
    pub current_step_id: Option<Uuid>,
    pub completed_step_count: i32,
    pub total_step_count: i32,
    pub sop_version_snapshot: SopVersionSnapshot,
    pub started_at: DateTime<Utc>,
    pub last_event_at: Option<DateTime<Utc>>,
    pub events: Vec<WorkEventSummary>,
}

// ─────────────────────────────────────────────────────────────────────────────
// API-work-execs-003: POST /api/v1/work-executions/{id}/suspend
// ─────────────────────────────────────────────────────────────────────────────

/// 作業中断リクエスト（API-work-execs-003）
#[derive(Debug, Deserialize, ToSchema)]
#[allow(dead_code)]
pub struct SuspendRequest {
    /// 中断理由コード（equipment_breakdown / material_shortage / quality_issue / emergency / other）
    pub reason_code: String,
    /// 補足説明（任意、最大 500 文字）
    pub reason_detail: Option<String>,
    /// クライアント側の中断時刻
    pub timestamp_client: DateTime<Utc>,
}

/// 作業中断レスポンスの data フィールド（API-work-execs-003）
#[derive(Debug, Serialize, ToSchema)]
pub struct SuspendData {
    pub id: Uuid,
    pub status: String,
    pub suspension_id: Uuid,
    pub suspended_at: DateTime<Utc>,
}

// ─────────────────────────────────────────────────────────────────────────────
// API-work-execs-004: POST /api/v1/work-executions/{id}/resume
// ─────────────────────────────────────────────────────────────────────────────

/// 作業再開リクエスト（API-work-execs-004）
#[derive(Debug, Deserialize, ToSchema)]
#[allow(dead_code)]
pub struct ResumeRequest {
    /// 再開実行者 ID（TBL-016 に存在すること）
    pub resumed_by: Uuid,
    /// クライアント側の再開時刻
    pub timestamp_client: DateTime<Utc>,
}

/// 作業再開レスポンスの data フィールド（API-work-execs-004）
#[derive(Debug, Serialize, ToSchema)]
pub struct ResumeData {
    pub id: Uuid,
    pub status: String,
    pub resumed_at: DateTime<Utc>,
    pub current_step_id: Option<Uuid>,
}

// ─────────────────────────────────────────────────────────────────────────────
// API-work-execs-005: POST /api/v1/work-executions/{id}/complete
// ─────────────────────────────────────────────────────────────────────────────

/// 作業完了リクエスト（API-work-execs-005）
#[derive(Debug, Deserialize, ToSchema)]
#[allow(dead_code)]
pub struct CompleteWorkRequest {
    /// 完了実行者 ID（TBL-016 に存在すること）
    pub completed_by: Uuid,
    /// クライアント側の完了時刻
    pub timestamp_client: DateTime<Utc>,
    /// 最終備考（任意、最大 1000 文字）
    pub final_remarks: Option<String>,
}

/// 作業完了レスポンスの data フィールド（API-work-execs-005）
#[derive(Debug, Serialize, ToSchema)]
pub struct CompleteWorkData {
    pub id: Uuid,
    pub status: String,
    pub completed_at: DateTime<Utc>,
    pub hash_chain_block_id: Option<Uuid>,
    pub hash_chain_value: Option<String>,
}
